util/codegen/codegen.go

```diff
@@ -38,3 +40,42 @@ func WriteFormatted(code []byte, path string) error {
 	}
 	return nil
 }
+
+// AssertStructUnchanged generates code that asserts at compile time that type t is unchanged.
+// tname is the named type corresponding to t.
+// ctx is a single-word context for this assertion, such as "Clone".
+// thisPkg is the package containing t.
+// If non-nil, AssertStructUnchanged will add elements to imports
+// for each package path that the caller must import for the returned code to compile.
+func AssertStructUnchanged(t *types.Struct, tname, ctx string, thisPkg *types.Package, imports map[string]struct{}) []byte {
```

Comment:
Optional:  Perhaps `thisPkg` should come right between `t` to match the fact that we usually refer to qualified type as `PkgPath.TypeName`

Reply:
Done. Did it as an additional commit to avoid pointless merge pain.